File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/notification/screen/NotificationViewModel.kt
package com.ssafy.ssabree.features.notification.screen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.NotificationRepository
import com.ssafy.ssabree.core.repository.model.NotificationModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class NotificationUiState(
    val isLoading: Boolean = false,
    val notifications: List<NotificationModel> = emptyList(),
    val error: String? = null
)

class NotificationViewModel(
    private val notificationRepository: NotificationRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(NotificationUiState())
    val uiState: StateFlow<NotificationUiState> = _uiState.asStateFlow()

    init {
        loadNotifications()
    }

    fun loadNotifications() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            notificationRepository.getNotifications()
                .onSuccess { notifications ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        notifications = notifications
                    )
                }
                .onFailure { e ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = e.message
                    )
                }
        }
    }

    fun markAsRead(id: Long) {
        viewModelScope.launch {
            notificationRepository.markAsRead(id)
                .onSuccess {
                    _uiState.value = _uiState.value.copy(
                        notifications = _uiState.value.notifications.map { notification ->
                            if (notification.id == id) {
                                notification.copy(isRead = true)
                            } else {
                                notification
                            }
                        }
                    )
                }
        }
    }
}

File: iOS/ssabree/ssabree/Features/Notification/NotificationViewModel.swift
import Foundation

// MARK: - UI State

struct NotificationUiState {
    var isLoading: Bool = false
    var notifications: [NotificationModel] = []
    var error: String? = nil
}

// MARK: - ViewModel

@Observable
@MainActor
final class NotificationViewModel {
    private let notificationRepository: NotificationRepository

    private(set) var uiState = NotificationUiState()

    init(notificationRepository: NotificationRepository) {
        self.notificationRepository = notificationRepository
    }

    func loadNotifications() async {
        uiState.isLoading = true
        uiState.error = nil

        let result = await notificationRepository.getNotifications()

        switch result {
        case .success(let notifications):
            uiState.notifications = notifications
            uiState.isLoading = false
        case .failure(let error):
            uiState.error = error.localizedDescription
            uiState.isLoading = false
        }
    }

    func markAsRead(id: Int) async {
        let result = await notificationRepository.markAsRead(id: id)

        if case .success = result {
            uiState.notifications = uiState.notifications.map { notification in
                if notification.id == id {
                    return notification.copyWith(isRead: true)
                }
                return notification
            }
        }
    }
}
